src/deepsparse/loggers/function_logger.py

```diff
@@ -0,0 +1,146 @@
+
+
+__all__ = ["FunctionLogger"]
+
+
+class FunctionLogger(BaseLogger):
+    """
+    DeepSparse logger that applies functions to raw log values
+    according to the specified config file
+
+    :param logger: A DeepSparse Logger object
+    :param config: A configuration dictionary that specifies the mapping between
+        the target name and the functions to be applied to raw log values
+
+        `config` can be specified as a dictionary type:
+        e.g.
+
+        {"pipeline_inputs":
+            [{"function": "some_function_1",
+              "frequency": 3},
+
+            [{"function": "some_function_2",
+              "frequency": 5}],
+
+         "pipeline_outputs":
+            [{"function": "some_function_1",
+            "frequency": 4}]
+        }
+
+        or a a yaml file:
+        e.g.
+
+        ```yaml
+        pipeline_inputs:
+            - function: some_function_1
+              frequency: 3
+            - function: some_function_2
+              frequency: 5
+        pipeline_outputs:
+            - function: some_function_1
+              frequency: 4
+        ```
+    """
+
+    def __init__(
+        self,
+        logger: BaseLogger,
+        config: Union[str, Dict[str, List[Dict[str, Any]]]],
+    ):
+
+        self.logger = logger
+        self.config = (
+            config
+            if isinstance(config, dict)
+            else yaml.safe_load(Path(config).read_text())
+        )
+        self.function_call_counter = self._create_function_call_counter(self.config)
+
+    def log(self, identifier: str, value: Any, category: MetricCategories):
+        """
+        :param identifier: The name of the thing that is being logged.
+        :param value: The data structure that the logger is logging
+        :param category: The metric category that the log belongs to
+        """
+        if category == MetricCategories.DATA:
+            # logging data information
+            pipeline_name, *target = identifier.split(".")
+            self._log_data(
+                pipeline_name=pipeline_name,
+                target=".".join(target),
+                value=value,
+                category=category,
+            )
+        else:
+            # logging system information
+            self.logger.log(
+                identifier=identifier,
+                value=value,
+                category=category,
+            )
+
+    def _log_data(
+        self,
+        pipeline_name: str,
+        target: str,
+        value: Any,
+        category: MetricCategories,
+    ):
+        list_target_functions = self.config.get(target)
+        if list_target_functions is None:
+            # no function to apply to the given target
+            return
+
+        for function_dict in list_target_functions:
+            function = function_dict.get("function")
+            logging_frequency = function_dict.get("frequency")
+
+            if self.function_call_counter[target][function] % logging_frequency == 0:
```

Comment:
What if the logging frequency is 0?

Reply:
Yeah, this should be prohibited, good catch. Adding a validation layer.

Reply:
Now:
```bash
pydantic.error_wrappers.ValidationError: 1 validation error for PipelineLoggingConfig
targets -> 0 -> mappings -> 0 -> frequency
  Passed frequency: 0, but frequency must be a positive integer greater equal 1 (type=value_error)
 ```